crates/fuel-core/src/service.rs

```diff
@@ -1,56 +1,78 @@
+use crate::{
+    database::Database,
+    service::adapters::P2PAdapter,
+};
+use fuel_core_services::{
+    RunnableService,
+    RunnableTask,
+    ServiceRunner,
+    State,
+    StateWatcher,
+};
 use std::{
     net::SocketAddr,
     panic,
 };
-use tokio::{
-    sync::oneshot,
-    task::JoinHandle,
-};
 use tracing::log::warn;
 
 pub use config::{
     Config,
     DbType,
     VMConfig,
 };
+pub use fuel_core_services::Service as ServiceTrait;
 
 pub mod adapters;
 pub mod config;
 pub(crate) mod genesis;
-pub mod graph_api;
 pub mod metrics;
-pub mod modules;
+pub mod sub_services;
 
-pub struct FuelService {
-    handle: JoinHandle<()>,
-    /// Shutdown the fuel service.
-    shutdown: oneshot::Sender<()>,
+#[derive(Clone)]
+pub struct SharedState {
+    /// The transaction pool shared state.
+    pub txpool: fuel_core_txpool::service::SharedState<P2PAdapter, Database>,
+    /// The P2P network shared state.
+    #[cfg(feature = "p2p")]
+    pub network: fuel_core_p2p::service::SharedState,
     #[cfg(feature = "relayer")]
-    /// Relayer handle
-    relayer_handle: Option<fuel_core_relayer::RelayerSynced>,
-    /// The address bound by the system for serving the API
-    pub bound_address: SocketAddr,
+    /// The Relayer shared state.
+    pub relayer: Option<fuel_core_relayer::SharedState>,
+    /// The GraphQL shared state.
+    pub graph_ql: crate::graph_api::service::SharedState,
 }
 
-struct FuelServiceInner {
-    tasks: Vec<JoinHandle<Result<(), AnyError>>>,
-    /// handler for all modules.
-    modules: Modules,
+pub struct FuelService {
+    /// The `ServiceRunner` used for `FuelService`.
+    ///
+    /// # Dev-note: The `FuelService` is already exposed as a public API and used by many crates.
+    /// To provide a user-friendly API and avoid breaking many downstream crates, `ServiceRunner`
+    /// is wrapped inside.
+    runner: ServiceRunner<Task>,
+    /// The shared state of the service
+    pub shared: SharedState,
     /// The address bound by the system for serving the API
     pub bound_address: SocketAddr,
-    /// Shutdown the graphql api
-    stop_graphql_api: oneshot::Sender<()>,
 }
 
 impl FuelService {
-    /// Create a fuel node instance from service config
-    #[tracing::instrument(skip(config))]
-    pub async fn new_node(mut config: Config) -> Result<Self, AnyError> {
+    /// Creates a `FuelService` instance from service config
+    pub fn new(database: Database, mut config: Config) -> anyhow::Result<Self> {
         Self::make_config_consistent(&mut config);
+        let task = Task::new(database, config)?;
+        let runner = ServiceRunner::new(task);
+        let shared = runner.shared.clone();
+        let bound_address = runner.shared.graph_ql.bound_address;
+        Ok(FuelService {
+            bound_address,
+            shared,
+            runner,
+        })
+    }
+
+    /// Creates and starts fuel node instance from service config
+    #[tracing::instrument(skip(config))]
```

Comment:
we should probably move this tracing annotation to the above `new` method so that tracing is always activated regardless of which entrypoint is used to start the svc.